controllers/image_controller.go

```diff
@@ -0,0 +1,255 @@
+			return cancel, imageSpecJob
+		}
+
+		// Rebuild if image phase was succeed or failed and image.updateTime after imageSpecJob.updateTime
+		imageUpdateTime := image.Spec.ImageSpec.UpdateTime
+		imageSpecJobUpdateTime := imageSpecJob.Spec.UpdateTime
+		if imageUpdateTime != nil && imageSpecJobUpdateTime != nil &&
+			imageUpdateTime.After(imageSpecJob.Spec.UpdateTime.Time) {
+			return rebuild, imageSpecJob
+		}
+
+		// Update if imageSpecJob exist
+		return update, imageSpecJob
+	}
+	return unknown, nil
+}
+
+func isImageCustomBuild(image *v1alpha1.Image) bool {
+	if image == nil || image.Spec.ImageSpec.BaseImage == "" {
+		return false
+	}
+	return true
+}
+
+func getImageSpecJobName(image *v1alpha1.Image) string {
+	if image == nil {
+		return ""
+	}
+	if image.Spec.ImageSpec.BaseImage == "" {
+		return ""
+	}
+	return image.Name
+}
+
+func createImageSpecJob(r *ImageReconciler, ctx context.Context, image *v1alpha1.Image) error {
+	if !isImageCustomBuild(image) {
+		return fmt.Errorf("creiateImageSpecJob do not provide correct image")
+	}
+
+	pushSecretName := viper.GetString("customImage.pushSecretName")
+	repoPrefix := viper.GetString("customImage.pushRepoPrefix")
+
+	hash := computeHash(image.Spec.ImageSpec.BaseImage, image.Spec.ImageSpec.Packages)
+
+	updateTime := image.Spec.ImageSpec.UpdateTime
+	if updateTime == nil {
+		updateTime = &metav1.Time{Time: time.Now()}
+	}
+
+	imageSpecJob := &v1alpha1.ImageSpecJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        getImageSpecJobName(image),
+			Namespace:   image.Namespace,
+			Annotations: map[string]string{"image.primehub.io/hash": hash},
+			Labels: map[string]string{
+				"image.primehub.io/name": image.ObjectMeta.Name,
+				"app":                    "primehub-image",
+			},
+		},
+		Spec: v1alpha1.ImageSpecJobSpec{
+			BaseImage:   image.Spec.ImageSpec.BaseImage,
+			PullSecret:  image.Spec.PullSecret,
+			Packages:    image.Spec.ImageSpec.Packages,
+			TargetImage: image.ObjectMeta.Name + ":" + hash,
+			PushSecret:  pushSecretName,
+			RepoPrefix:  repoPrefix,
+			UpdateTime:  updateTime,
+		},
+	}
+
+	if err := ctrl.SetControllerReference(image, imageSpecJob, r.Scheme); err != nil {
+		return err
+	}
+	if err := r.Client.Create(ctx, imageSpecJob); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func cancelImageSpecJob(r *ImageReconciler, ctx context.Context, image *v1alpha1.Image, imageSpecJob *v1alpha1.ImageSpecJob) error {
+	if !isImageCustomBuild(image) {
+		return fmt.Errorf("cancelImageSpecJob do not provide correct image")
+	}
+	if imageSpecJob == nil {
+		return fmt.Errorf("cancelImageSpecJob do not provide correct imageSpecJob")
+	}
+
+	if err := r.Delete(ctx, imageSpecJob, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+		return err
+	}
+
+	imageClone := image.DeepCopy()
+	imageClone.Status.JobCondiction.Phase = CustomImageJobStatusCancelled
```

Comment:
Currently, the cancelled phase would be overwrite. Could we always mark the Phase as `Cancelled` if the `spec.imageSpec.cancel` is true?

Reply:
Add logic to check if the image phase is not cancelled but the image had been marked cancelled.